### DISCUSSION TITLE
Escape codes for links aren't always suppressed when not supported

### DISCUSSION THREAD

**Discussion by emojjon** [Textualize/rich#1625] [General]
As I understand it, you can't create an instance of Console that explicitly doesn't use link escape codes. Apparently these codes should be ignored by terminals that don't support them. Sometimes it would be good though, because the codes actually mangles the output in some circumstances.

I hesitate to say exactly how to reproduce this since in my case it involved several ssh logins and strangeness with shell rc-files and what not. Suffice to say it would be nice to be able to turn it off in some circumstances, perhaps even from outside the python code (environment variable?).

Great project by the way. I discovered it yesterday, and I'm looking forward to try and use it myself.

**Comment by emojjon** [comment]
Here are two images to illustrate the problem with link codes. To test it I added import os and changed the corresponding line (around 706 I think) in style.py to:

if 'SUPPORTS_LINKS' in os.environ and os.environ['SUPPORTS_LINKS'] and self._link and not legacy_windows

I expect you might want to do something a bit more elegant, but I think it illustrates my point that it might be good to be able to turn this feature off. Probably at different levels. Ideally one would of course like to have a reliable test if these links can be handled (or at least ignored) or not.

![WithLinkCodes](https://user-images.githubusercontent.com/66027466/139317674-ad12ca95-5c79-4f71-9c9e-45c351986cf4.png)
![WithoutLinkCodes](https://user-images.githubusercontent.com/66027466/139317677-7af45a24-18ff-48e0-8f7d-21fd40f73437.png)
:

**Comment by willmcgugan** [comment]
The link code should be invisible if not supported by the terminal, as defined by the spec. If they are visible, then the terminal has broken behaviour.

That said, if there is a terminal with that broken behaviour in common use, I would want to add a way of disabling them.

But I'm going to need a way of reproducing the issue. It's entirely possible that the fault is not the terminal, and its some intermediate software which is misinterpreting these codes and making them visible.

Can tell me how to reproduce this issue?

**Comment by emojjon** [comment]
I am hesitant if I can figure out how to reproduce it because the actual terminal was one terminal running here with quite a lot of  local scripts connecting to another computer, with it's own setup. Possibly I might even have ssh'ed to one computer and on to the third. "tmux" might have been involved. I don't have the time to do this right now but I might get back to you. I do think that you should have some small mechanism to turn this off anyway for those people who find themselves working with such a terminal environment simply because implementing  it would be extremely simple. Just mention somewhere how to turn the links off and put in an extra conditional in the "render" function in style.py to check if it should include the link code. In fact you already have such an exception for "legacy_windows" but it might not be so easy for users to set that variable (and also wrong, because I'm sure that setting it has other effects too).

Anyway, do with this what you will! I might get back to you if I stumble over a good way to reproduce the problem. In the mean time I've jerry-rigged the code that I use to get around the problem.

/Johan
________________________________

**Comment by willmcgugan** [comment]
I suspect that it was an issue with tmux. I've just tried and tmux doesn't display links, even though the terminal (iTerm) will render them. Which means that tmux is stripping out those codes.

I wouldn't be surprised if you happened to be running an older version of tmux which didn't know how to handle those link codes.

There are a few options, a env var like you suggest, or I could detect tmux. But I'd prefer to wait until I knew for certain why this was happening.

Let me know if you figure something out.

**Comment by emojjon** [comment]
If and when I have time I might figure out when this happens. I think you should put in a simple fix like an env. var. simply because it costs so little (the normal behaviour will be the default anyway). There probably is a "complete" solution but it might take some work so it probably would have to wait for more prioritized things.

Can't tmux be fairly confidently detected by checking for the 'TMUX' env. variable? Admittedly you could set it outside tmux and probably unset it inside tmux too, but then one has probably made a conscious choice of which one is aware.

Keep up the good work!

/Johan
________________________________
